File: bench/s03_optional_result.c
/* s03_optional_result.c — Optional/Result hot-path overhead
 * Compares "direct return" vs "optional with ok-flag" function call overhead.
 * N=100000000 calls. compute(x) = x * 42 + 7.
 * Direct:   int32_t compute_direct(int32_t x) { return x*42 + 7; }
 * Optional: int32_t compute_optional(int32_t x, int32_t *ok) { if (x > 1000000) { *ok=0; return 0; } *ok=1; return x*42+7; }
 * Anti-fold: __asm__ volatile in bench loop bodies; bench returns sum. */
#include <stdint.h>
#include <stdio.h>
#include <time.h>

enum { N = 100000000 };

static int32_t compute_direct(int32_t x) {
  return x * 42 + 7;
}

static int32_t compute_optional(int32_t x, int32_t *ok) {
  if (x > 1000000) {
    *ok = 0;
    return 0;
  }
  *ok = 1;
  return x * 42 + 7;
}

static int32_t bench_direct(void) {
  int32_t sum = 0;
  int32_t i = 0;
  while (i < N) {
    sum += compute_direct(i);
    i = i + 1;
    __asm__ volatile("" : "+r"(sum) : : "memory");
  }
  return sum;
}

static int32_t bench_optional(void) {
  int32_t sum = 0;
  int32_t i = 0;
  while (i < N) {
    int32_t ok;
    int32_t v = compute_optional(i, &ok);
    if (ok) {
      sum += v;
    }
    i = i + 1;
    __asm__ volatile("" : "+r"(sum) : : "memory");
  }
  return sum;
}

int main(void) {
  clock_t t1 = clock();
  int32_t s1 = bench_direct();
  clock_t t2 = clock();
  int32_t s2 = bench_optional();
  clock_t t3 = clock();

  double ms1 = (double)(t2 - t1) / CLOCKS_PER_SEC * 1000.0;
  double ms2 = (double)(t3 - t2) / CLOCKS_PER_SEC * 1000.0;
  printf("direct:    %8.2f ms  sum=%d\n", ms1, (int)s1);
  printf("optional:  %8.2f ms  sum=%d\n", ms2, (int)s2);
  return 0;
}

File: bench/s03_optional_result.zig
// s03_optional_result.zig — Optional/Result hot-path overhead
// Compares "direct return" vs "optional (?i32)" function call overhead.
// Matches bench/s03_optional_result.c / .x algorithm.
// N=100000000 calls. compute(x) = x * 42 + 7.
// Optional variant uses ?i32 with null on failure (x > 1000000).
const std = @import("std");

const N: i32 = 100_000_000;

fn compute_direct(x: i32) i32 {
    return x *% 42 +% 7;
}

fn compute_optional(x: i32) ?i32 {
    if (x > 1_000_000) return null;
    return x *% 42 +% 7;
}

fn bench_direct() i32 {
    var sum: i32 = 0;
    var i: i32 = 0;
    while (i < N) : (i += 1) {
        sum +%= compute_direct(i);
    }
    return sum;
}

fn bench_optional() i32 {
    var sum: i32 = 0;
    var i: i32 = 0;
    while (i < N) : (i += 1) {
        if (compute_optional(i)) |v| {
            sum +%= v;
        }
    }
    return sum;
}

pub fn main(init: std.process.Init) !void {
    const ts0 = std.Io.Timestamp.now(init.io, .awake);
    const s1 = bench_direct();
    const ts1 = std.Io.Timestamp.now(init.io, .awake);
    const s2 = bench_optional();
    const ts2 = std.Io.Timestamp.now(init.io, .awake);

    const ns1: i64 = @intCast(ts0.durationTo(ts1).nanoseconds);
    const ns2: i64 = @intCast(ts1.durationTo(ts2).nanoseconds);

    const ms1 = @as(f64, @floatFromInt(ns1)) / 1_000_000.0;
    const ms2 = @as(f64, @floatFromInt(ns2)) / 1_000_000.0;

    std.debug.print("direct:    {d:.2} ms  sum={d}\n", .{ ms1, s1 });
    std.debug.print("optional:  {d:.2} ms  sum={d}\n", .{ ms2, s2 });
}
